src/huggingface/weight.jl

```diff
@@ -14,18 +15,30 @@ Load the `state_dict` from the given `model_name` from huggingface hub. By defau
 """
 function load_state_dict(model_name; possible_files = nothing, kw...)
     possible_files = ensure_possible_files(possible_files, model_name; kw...)
+    # here we need to create a fork to recognize, if we want to load the weights from safetensors
+    # or from pytorch pickle
     if PYTORCH_WEIGHTS_INDEX_NAME in possible_files
-        weight_index = JSON3.read(read(hgf_model_weight_index(model_name; kw...)))
-        full_state_dict = OrderedDict{Any, Any}()
-        for weight_file in Set{String}(values(weight_index.weight_map))
-            merge!(full_state_dict, Pickle.Torch.THload(hgf_file(model_name, weight_file; kw...)))
-        end
-        return full_state_dict
+      weight_index = JSON3.read(read(hgf_model_weight_index(model_name; kw...)))
+      return load_weights_from_weightmap(Pickle.Torch.THload, model_name, weight_index;kw...)
+    elseif SAFETENSORS_WEIGHTS_INDEX_NAME in possible_files
+      weight_index = JSON3.read(read(hgf_model_safetensor_index(model_name; kw...)))
+      return load_weights_from_weightmap(
+    elseif any(s -> endswith(s, "safetensors"), possible_files)
+      error("seems like weights are stored in safetensors, but the loader is not implemented since the author did not know the name of the file. File issue with name of the repo and tag @pevnak")
+      # return load_safetensors(hgf_model_safetensor(model_name; kw...))
```

Comment:
A small concern about the safetensors format is that it does not support weight sharing. According to the python implementation, they have handler in the `transformers` code that only extracts a unique weight if any duplication exists. It's unclear whether we also need a handler. It depends on how they pick the unique weight because our loader checks if a specific weight name exists.

Reply:
I can add a manual deduplication, e.g. if the L0 distance between two tensors is smaller then 1e-5, we can consider them equal.

Reply:
Do you have some link with that? I can imagine that JSON will have two entries pointing to the same area in the file, which can be easily detected and weights can be shared.

Reply:
I do not know, if to put it to the pr. It seems to me to be the problem of SafeTensors loader

Reply:
For reference, they give [reasons for not supporting weight sharing in the doc](https://huggingface.co/docs/safetensors/en/torch_shared_tensors#why-are-shared-tensors-not-saved-in-safetensors-). Thus, weight sharing is handled at different level, see [this function](https://github.com/huggingface/safetensors/blob/b947b59079a6197d7930dfb535818ac4896113e8/bindings/python/py_src/safetensors/torch.py#L84) and its usage in `load_model`/`save_model` in the same file.

Reply:
According to their `save_model`, they might record the shared tensor key in the metadata. I made a patch that would try to recover the missing key from the metadata (if present) and log any missing/extra keys. This is probably the only strategy we can do for now. Would also need https://github.com/FluxML/SafeTensors.jl/pull/5